src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -0,0 +1,593 @@
+        }
+
+        // Walk through, merging flownodes one-by-one until everything has merged to one ancestor
+        while (iterators.size() > 1) {
+            ListIterator<Filterator<FlowNode>> itIterator = iterators.listIterator();
+            ListIterator<FlowPiece> pieceIterator = livePieces.listIterator();
+
+            while (itIterator.hasNext()) {
+                Filterator<FlowNode> blockStartIterator = itIterator.next();
+                FlowPiece myPiece = pieceIterator.next(); //Safe because we always remove/add with both iterators at once
+
+                // Welp we hit the end of a branch
+                if (!blockStartIterator.hasNext()) {
+                    pieceIterator.remove();
+                    itIterator.remove();
+                    continue;
+                }
+
+                FlowNode nextBlockStart = blockStartIterator.next();
+
+                // Look for cases where two branches merge together
+                FlowPiece existingPiece = branches.get(nextBlockStart);
+                if (existingPiece == null && myPiece instanceof FlowSegment) { // No merge, just add to segment
+                    ((FlowSegment) myPiece).add(nextBlockStart);
+                    branches.put(nextBlockStart, myPiece);
+                } else if (existingPiece == null && myPiece instanceof Fork) {  // No merge, we had a fork. Start a segment preceding the fork
+                    FlowSegment newSegment = new FlowSegment();
+                    newSegment.isLeaf = false;
+                    newSegment.add(nextBlockStart);
+                    newSegment.after = myPiece;
+                    pieceIterator.remove();
+                    pieceIterator.add(newSegment);
+                    branches.put(nextBlockStart, newSegment);
+                } else if (existingPiece != null) {  // Always not null. We're merging into another thing, we're going to elliminate a branch
+                    if (existingPiece instanceof Fork) {
+                        ((Fork) existingPiece).following.add(myPiece);
+                    } else { // Split a flow segment so it forks against this one
+                        Fork f = ((FlowSegment) existingPiece).split(branches, (BlockStartNode)nextBlockStart, myPiece);
+                        // If we split the existing segment at its end, we created a fork replacing its latest node
+                        // Thus we must replace the piece with the fork ahead of it
+                        if (f.following.contains(existingPiece) ) {
+                            int headIndex = livePieces.indexOf(existingPiece);
+                            livePieces.set(headIndex, f);
+                        }
+                        parallelForks.add(f);
+                    }
+
+                    // Merging removes the piece & its iterator from heads
+                    itIterator.remove();
+                    pieceIterator.remove();
+                }
+            }
+        }
```

Comment:
🐜 Generally this logic assumes that FlowPiece is either Fork or FlowSegment. In the case of other items we get into ClassCastException or undefined behavior. 

The FlowNode interface is not public, so I think it's actually NIT. But looks weird